src/cpp/fastdds/subscriber/DataReaderImpl.cpp

```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
+        max_samples = qos_.reader_resource_limits().max_samples_per_read;
+    }
+
+    return ReturnCode_t::RETCODE_OK;
+}
+
+ReturnCode_t DataReaderImpl::prepare_loan(
+        LoanableCollection& data_values,
+        SampleInfoSeq& sample_infos,
+        int32_t& max_samples)
+{
+    if (0 < data_values.maximum())
+    {
+        // A loan was not requested
+        return ReturnCode_t::RETCODE_OK;
+    }
+
+    if (max_samples > 0)
+    {
+        // Check if there are enough sample_infos
+        size_t num_infos = sample_info_pool_.num_allocated();
+        if (num_infos == qos_.reader_resource_limits().sample_infos_allocation.maximum)
+        {
+            return ReturnCode_t::RETCODE_OUT_OF_RESOURCES;
+        }
+
+        // Limit max_samples to available sample_infos
+        num_infos += max_samples;
+        if (num_infos > qos_.reader_resource_limits().sample_infos_allocation.maximum)
+        {
+            size_t exceed = num_infos - qos_.reader_resource_limits().sample_infos_allocation.maximum;
+            max_samples -= static_cast<uint32_t>(exceed);
+        }
+    }
+
+    // Check if there are enough loans
+    ReturnCode_t code = loan_manager_.get_loan(data_values, sample_infos);
+    if (!code)
+    {
+        return code;
+    }
+
+    return ReturnCode_t::RETCODE_OK;
+}
+
+ReturnCode_t DataReaderImpl::read(
+        LoanableCollection& data_values,
+        SampleInfoSeq& sample_infos,
+        int32_t max_samples,
+        SampleStateMask sample_states,
+        ViewStateMask view_states,
+        InstanceStateMask instance_states)
+{
+    if (reader_ == nullptr)
+    {
+        return ReturnCode_t::RETCODE_NOT_ENABLED;
+    }
+
+    ReturnCode_t code = check_collection_preconditions_and_calc_max_samples(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
+    }
+
+    std::lock_guard<RecursiveTimedMutex> lock(reader_->getMutex());
+
+    code = prepare_loan(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
+    }
+
+    auto instance = history_.lookup_instance(HANDLE_NIL, false).second;
+    detail::StateFilter states{ sample_states, view_states, instance_states };
+    detail::ReadTakeCommand cmd(*this, data_values, sample_infos, max_samples, states, instance);
+    while (!cmd.is_finished())
+    {
+        cmd.add_instance(false);
+    }
+    return cmd.return_value();
+}
+
+ReturnCode_t DataReaderImpl::read_instance(
+        LoanableCollection& data_values,
+        SampleInfoSeq& sample_infos,
+        int32_t max_samples,
+        const InstanceHandle_t& a_handle,
+        SampleStateMask sample_states,
+        ViewStateMask view_states,
+        InstanceStateMask instance_states)
+{
+    if (reader_ == nullptr)
+    {
+        return ReturnCode_t::RETCODE_NOT_ENABLED;
+    }
+
+    ReturnCode_t code = check_collection_preconditions_and_calc_max_samples(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
```

Comment:
Missing negative test

Reply:
This should be checked by `collection_preconditions` test.